scarb/src/ops/fmt.rs

```diff
@@ -0,0 +1,182 @@
+        let entry_path = dir_entry.unwrap();
+        if entry_path.file_type().unwrap().is_dir() {
+            return Continue;
+        }
+        let path = entry_path.path();
+        info!("Formatting file: {}.", path.display());
+        if self.opts.check {
+            match self.fmt.check(&path) {
+                Ok(FormatOutcome::Identical(_)) => {}
+                Ok(FormatOutcome::DiffFound(diff)) => {
+                    if self.opts.color {
+                        self.ws.config().ui().print(format!(
+                            "Diff found in file {}:\n {}",
+                            path.display(),
+                            diff
+                        ));
+                    } else {
+                        self.ws.config().ui().print(format!(
+                            "Diff found in file {}:\n {}",
+                            path.display(),
+                            create_patch(&diff.original, &diff.formatted)
+                        ));
+                    };
+
+                    self.own_correct = false;
+                }
+                Err(parsing_error) => {
+                    print_error(parsing_error, path.display().to_string(), self.ws);
+                    self.own_correct = false;
+                }
+            }
+        } else if let Err(parsing_error) = self.fmt.format_in_place(&path) {
+            print_error(parsing_error, path.display().to_string(), self.ws);
+            self.own_correct = false;
+        }
+        Continue
+    }
+}
+
+impl<'t> Drop for PathFormatter<'t> {
+    fn drop(&mut self) {
+        *self.all_correct.lock().unwrap() &= self.own_correct;
+    }
+}
+
+fn do_format_package(opts: &FmtOptions, pkg: &Package, ws: &Workspace<'_>) -> Result<bool> {
+    let config = FormatterConfig::default();
+    let fmt = CairoFormatter::new(config);
+
+    let base_path = Path::new(pkg.manifest_path().parent().unwrap());
+    let walk = fmt.walk(base_path);
+    let all_correct = Mutex::new(true);
+    let mut builder = PathFormatterBuilder {
+        ws,
+        opts,
+        fmt: &fmt,
+        all_correct: &all_correct,
+    };
+    walk.build_parallel().visit(&mut builder);
+    let result = *builder.all_correct.lock().unwrap();
+
+    Ok(result)
+}
+
+fn format_package(opts: &FmtOptions, pkg: &Package, ws: &Workspace<'_>) -> Result<()> {
+    ws.config()
+        .ui()
+        .print(Status::new("Formatting", "green", &pkg.id.to_string()));
+    if !do_format_package(opts, pkg, ws)? {
+        exit_fail(ws);
+    }
+    Ok(())
+}
+
+fn exit_fail(ws: &Workspace<'_>) -> ! {
+    let elapsed_time = HumanDuration(ws.config().elapsed_time());
+    ws.config().ui().print(Status::new(
+        "Finished with error",
+        "red",
+        &format!("in {elapsed_time}"),
+    ));
+    std::process::exit(1);
+}
+
+fn format_package_by_condition<F: Fn(&Package) -> bool>(
+    ws: &Workspace<'_>,
+    opts: &FmtOptions,
+    cond: F,
+) -> Result<()> {
```

Comment:
```suggestion
fn format_package_by_condition(
    ws: &Workspace<'_>,
    opts: &FmtOptions,
    cond: &dyn Fn(&Package) -> bool,
) -> Result<()> {
```

You use this function in three places with three distinct function types (each function/closure creates a distinct function type, all of them implement either `Fn`, `FnMut` or `FnOnce` trait). This causes Rust to create 3 copies of `format_package_by_condition` during monomorphisation, which is really suboptimal. Using trait objects is fine here, this is not a hot path.